src/providers/swift.rs

```diff
@@ -0,0 +1,176 @@
+use super::Provider;
+use crate::nixpacks::{
+    app::App,
+    environment::Environment,
+    nix::pkg::Pkg,
+    phase::{BuildPhase, InstallPhase, SetupPhase, StartPhase},
+};
+use anyhow::{bail, Result};
+
+static DEFAULT_SWIFT_VERSION: &str = "5.5.3";
+
+// From: https://lazamar.co.uk/nix-versions/?channel=nixpkgs-unstable&package=swift
+static AVAILABLE_SWIFT_VERSIONS: &[(&str, &str)] = &[
+    ("3.1", "aeaa79dc82980869a88a5955ea3cd3e1944b7d80"),
+    ("3.1.1", "8414d8386b9a6b855b291fb3f01a4e3b04c08bbb"),
+    ("4.0.3", "2c9d2d65266c2c3aca1e4c80215de8bee5295b04"),
+    ("4.1", "92a047a6c4d46a222e9c323ea85882d0a7a13af8"),
+    ("4.1.3", "a3962299f14944a0e9ccf8fd84bd7be524b74cd6"),
+    ("4.2.1", "7ff8a16f0726342f0a25697867d8c1306d4da7b0"),
+    ("4.2.3", "3fa154fd7fed3d6a94322bf08a6def47d6f8e0f6"),
+    ("5.0.1", "4599f2bb9a5a6b1482e72521ead95cb24e0aa819"),
+    ("5.0.2", "a9eb3eed170fa916e0a8364e5227ee661af76fde"),
+    ("5.1.1", "9986226d5182c368b7be1db1ab2f7488508b5a87"),
+    ("5.4", "c82b46413401efa740a0b994f52e9903a4f6dcd5"),
+    ("5.4.2", "c82b46413401efa740a0b994f52e9903a4f6dcd5"),
+    ("5.5.2", "6d02a514db95d3179f001a5a204595f17b89cb32"),
+];
+
+pub struct SwiftProvider {}
+
+impl Provider for SwiftProvider {
+    fn name(&self) -> &str {
+        "swift"
+    }
+
+    fn detect(&self, app: &App, _env: &Environment) -> Result<bool> {
+        Ok(app.includes_file("Package.swift")
+            || (app.includes_file("Package.swift") && app.includes_file("Package.resolved")))
+    }
+
+    fn setup(&self, app: &App, _env: &Environment) -> Result<Option<SetupPhase>> {
+        let mut setup_phase = SetupPhase::new(vec![
+            Pkg::new("coreutils"),
+            Pkg::new("swift"),
+            Pkg::new("clang"),
+            Pkg::new("zlib"),
+            Pkg::new("zlib.dev"),
+        ]);
+
+        let swift_version = SwiftProvider::get_swift_version(app)?;
+        let rev = version_number_to_rev(&swift_version)?;
+
+        if let Some(rev) = rev {
+            setup_phase.set_archive(rev);
+        }
+
+        Ok(Some(setup_phase))
+    }
+
+    fn install(&self, app: &App, _env: &Environment) -> Result<Option<InstallPhase>> {
+        let mut install_phase = InstallPhase::new("swift package resolve".to_string());
+
+        install_phase.add_file_dependency("Package.swift".to_string());
+
+        if app.includes_file("Package.resolved") {
+            install_phase.add_file_dependency("Package.resolved".to_string());
+        }
+
+        Ok(Some(install_phase))
+    }
+
+    fn build(&self, app: &App, _env: &Environment) -> Result<Option<BuildPhase>> {
+        let name = SwiftProvider::get_executable_name(app)?;
+        let build_cmd = format!(
+            "\
+        CC=clang++ \
+        CPATH=~/.nix-profile/include \
+        LIBRARY_PATH=~/.nix-profile/lib \
+        QTDIR=~/.nix-profile \
```

Comment:
Variables should be handled via `environment_variables` function. Example: [Node provider](https://github.com/railwayapp/nixpacks/blob/7d8f3d718f7c0e75e20b2a51e7496c6048d71d78/src/providers/node.rs#L67-L73)

Reply:
This is just for the build & the `clang` linker tho. Setting them globally may cause possible unwanted behavior.

Reply:
What sort of unwanted behaviors? Mind providing an example where this can be reproduced?

Reply:
Nevermind then. The `CPATH`, `LIBRARY_PATH`, and the `QTDIR` variables should be probably set in the base image.

ref: https://nixos.wiki/wiki/Nix_vs._Linux_Standard_Base#Build_and_install_from_Source